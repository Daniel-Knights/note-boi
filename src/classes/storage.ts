import { DeletedNote } from '../api';
import { ColourTheme, UpdateStrategy } from '../constant';

export const STORAGE_KEYS_STRING = {
  MENU_WIDTH: 'note-menu-width',
  THEME: 'theme',
  UPDATE_SEEN: 'update-seen',
  UPDATE_STRATEGY: 'update-strategy',
  USERNAME: 'username',
} satisfies Record<StorageKeyString, string>;

export const STORAGE_KEYS_JSON = {
  UNSYNCED: 'unsynced-note-uuids',
} satisfies Record<StorageKeyJson, string>;

export const STORAGE_KEYS = { ...STORAGE_KEYS_STRING, ...STORAGE_KEYS_JSON };

/** `localStorage` convenience methods with stronger typing. */
export class Storage {
  static get<T extends keyof StorageValuesString>(key: T): StorageValues[T] | null {
    return localStorage.getItem(STORAGE_KEYS[key]) as StorageValues[T] | null;
  }

  static set<T extends keyof StorageValuesString>(key: T, value: StorageValues[T]): void {
    return localStorage.setItem(STORAGE_KEYS[key], value as string);
  }

  static remove(key: StorageKey) {
    return localStorage.removeItem(STORAGE_KEYS[key]);
  }

  static clear() {
    return localStorage.clear();
  }

  static getJson<T extends keyof StorageValuesJson>(key: T): StorageValues[T] | null {
    const storedItem = localStorage.getItem(STORAGE_KEYS[key]);

    return storedItem ? JSON.parse(storedItem) : null;
  }

  static setJson<T extends keyof StorageValuesJson>(key: T, value: StorageValues[T]) {
    return localStorage.setItem(STORAGE_KEYS[key], JSON.stringify(value));
  }
}

//// Types

type StorageValues = {
  MENU_WIDTH: string;
  THEME: ColourTheme;
  UNSYNCED: {
    new: string;
    edited: string[];
    deleted: DeletedNote[];
  };
  UPDATE_SEEN: string;
  UPDATE_STRATEGY: UpdateStrategy;
  USERNAME: string;
};

export type StorageKey = keyof StorageValues;

type StorageValuesString = {
  [K in StorageKey as StorageValues[K] extends string ? K : never]: StorageValues[K];
};

type StorageValuesJson = {
  [K in StorageKey as StorageValues[K] extends object ? K : never]: StorageValues[K];
};

export type StorageKeyString = keyof StorageValuesString;
export type StorageKeyJson = keyof StorageValuesJson;
