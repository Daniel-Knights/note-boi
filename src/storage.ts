import { ColourTheme, UpdateStrategy } from './constant';

export const STORAGE_KEYS_STRING = {
  USERNAME: 'username',
  THEME: 'theme',
  MENU_WIDTH: 'note-menu-width',
  UPDATE_SEEN: 'update-seen',
  UPDATE_STRATEGY: 'update-strategy',
} as const;

export const STORAGE_KEYS_JSON = {
  UNSYNCED: 'unsynced-note-ids',
} as const;

export const STORAGE_KEYS = { ...STORAGE_KEYS_STRING, ...STORAGE_KEYS_JSON };

export type StorageKey = keyof typeof STORAGE_KEYS;
export type StorageKeyString = keyof typeof STORAGE_KEYS_STRING;
export type StorageKeyJson = keyof typeof STORAGE_KEYS_JSON;

type StorageValues = {
  USERNAME: string;
  UNSYNCED: {
    new: string;
    edited: string[];
    deleted: string[];
  };
  THEME: ColourTheme;
  MENU_WIDTH: string;
  UPDATE_SEEN: string;
  UPDATE_STRATEGY: UpdateStrategy;
};

type StorageValuesString = {
  [K in StorageKey as StorageValues[K] extends string ? K : never]: StorageValues[K];
};

type StorageValuesJson = {
  [K in StorageKey as StorageValues[K] extends object ? K : never]: StorageValues[K];
};

/** `localStorage` convenience methods with stronger typing. */
export const storage = {
  get<T extends keyof StorageValuesString>(key: T): StorageValues[T] | null {
    return localStorage.getItem(STORAGE_KEYS[key]) as StorageValues[T] | null;
  },

  set<T extends keyof StorageValuesString>(key: T, value: StorageValues[T]): void {
    return localStorage.setItem(STORAGE_KEYS[key], value as string);
  },

  remove(key: StorageKey) {
    return localStorage.removeItem(STORAGE_KEYS[key]);
  },

  clear() {
    return localStorage.clear();
  },

  getJson<T extends keyof StorageValuesJson>(key: T): StorageValues[T] | null {
    const storedItem = localStorage.getItem(STORAGE_KEYS[key]);

    return storedItem ? JSON.parse(storedItem) : null;
  },

  setJson<T extends keyof StorageValuesJson>(key: T, value: StorageValues[T]) {
    return localStorage.setItem(STORAGE_KEYS[key], JSON.stringify(value));
  },
};
