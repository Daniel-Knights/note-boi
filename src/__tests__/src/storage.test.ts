import { storage } from '../../storage';

describe('storage', () => {
  it('get/set', () => {
    assert.isNull(storage.get('USERNAME'));

    storage.set('USERNAME', 'd');

    assert.strictEqual(storage.get('USERNAME'), 'd');
  });

  it('remove', () => {
    storage.set('USERNAME', 'd');

    storage.remove('USERNAME');

    assert.isNull(storage.get('USERNAME'));
  });

  it('clear', () => {
    storage.set('USERNAME', 'd');
    storage.set('THEME', 'Dark');

    storage.clear();

    assert.isNull(storage.get('USERNAME'));
    assert.isNull(storage.get('THEME'));
  });

  it('setJson/getJson', () => {
    assert.isNull(storage.getJson('UNSYNCED'));

    const unsynced = { new: 'note1', edited: ['note2'], deleted: ['note3'] };

    storage.setJson('UNSYNCED', unsynced);

    assert.deepEqual(storage.getJson('UNSYNCED'), unsynced);
  });
});
