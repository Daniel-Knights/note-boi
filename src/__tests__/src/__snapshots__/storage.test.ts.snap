// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`storage > clear > KeyStore.getKey() 1`] = `undefined`;

exports[`storage > clear > allCalls 1`] = `[]`;

exports[`storage > clear > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`storage > clear > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`storage > clear > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`storage > clear > openedPopup 1`] = `undefined`;

exports[`storage > clear > selectedTheme 1`] = `"System"`;

exports[`storage > clear > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`storage > clear > storage.get('THEME') 1`] = `null`;

exports[`storage > clear > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`storage > clear > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`storage > clear > storage.get('USERNAME') 1`] = `null`;

exports[`storage > clear > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`storage > clear > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`storage > get/set > KeyStore.getKey() 1`] = `undefined`;

exports[`storage > get/set > allCalls 1`] = `[]`;

exports[`storage > get/set > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`storage > get/set > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`storage > get/set > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`storage > get/set > openedPopup 1`] = `undefined`;

exports[`storage > get/set > selectedTheme 1`] = `"System"`;

exports[`storage > get/set > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`storage > get/set > storage.get('THEME') 1`] = `null`;

exports[`storage > get/set > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`storage > get/set > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`storage > get/set > storage.get('USERNAME') 1`] = `"d"`;

exports[`storage > get/set > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`storage > get/set > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`storage > remove > KeyStore.getKey() 1`] = `undefined`;

exports[`storage > remove > allCalls 1`] = `[]`;

exports[`storage > remove > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`storage > remove > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`storage > remove > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`storage > remove > openedPopup 1`] = `undefined`;

exports[`storage > remove > selectedTheme 1`] = `"System"`;

exports[`storage > remove > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`storage > remove > storage.get('THEME') 1`] = `null`;

exports[`storage > remove > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`storage > remove > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`storage > remove > storage.get('USERNAME') 1`] = `null`;

exports[`storage > remove > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`storage > remove > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`storage > setJson/getJson > KeyStore.getKey() 1`] = `undefined`;

exports[`storage > setJson/getJson > allCalls 1`] = `[]`;

exports[`storage > setJson/getJson > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`storage > setJson/getJson > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`storage > setJson/getJson > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`storage > setJson/getJson > openedPopup 1`] = `undefined`;

exports[`storage > setJson/getJson > selectedTheme 1`] = `"System"`;

exports[`storage > setJson/getJson > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`storage > setJson/getJson > storage.get('THEME') 1`] = `null`;

exports[`storage > setJson/getJson > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`storage > setJson/getJson > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`storage > setJson/getJson > storage.get('USERNAME') 1`] = `null`;

exports[`storage > setJson/getJson > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`storage > setJson/getJson > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
